archive_upload/handlers/dsmc_handlers.py

```diff
@@ -0,0 +1,754 @@
+
+        log.debug("Uploaded files to PDC: {}".format(matched_lines))
+
+        uploaded_files = {}
+
+        # We need to convert the sizes to a common format for easier comparison with local size.
+        # An output line can look like
+        #4,096  B  2017-07-27 17.48.34    /data/mm-xart002/runfolders/johanhe_test_0809_001-AG2UJ_archive/Config Never e374bd6b-ab36-4f41-94d3-f4eaea9f30d4
+        # but varies, depending on the environment's locale. Can e.g. be "4 096",
+        for line in matched_lines:
+            elements = line.split(" B ")
+            size = elements[0].strip()
+
+            if "," in size:
+                byte_size = size.replace(",", "")
+            elif " " in size:
+                byte_size = size.replace(" ", "")
+            else:
+                byte_size = size
+
+            # We can't be completely sure what format the timestamp will be returned with.
+            # And we can not be 100% sure what format the description will have either, at least in the future.
+            # This will have to do for now.
+            import re
+            substr = re.search("{}(.*) Never ".format(path_to_archive), line)
+            filename = ("{}{}".format(path_to_archive, substr.group(1))).strip()
+
+            # NB A potential error here is if the same file has been uploaded multiple times with the same descriptions.
+            # It is then a bit ambigious what to do. TSM sorts and returns them in chronological order though,
+            # so we will just keep refering to the last uploaded version of the file.
+            if filename in uploaded_files.keys():
+                log.info("Duplicate uploads of file {} with description {} encountered.".format(filename, descr))
+
+            uploaded_files[filename] = int(byte_size)
+
+        log.debug("Previously uploaded files for the archive are: {}".format(uploaded_files))
+
+        return uploaded_files
+
+    def get_local_filelist(self, path_to_archive):
+        """
+        Gets the list of all files and their sizes in the local archive.
+
+        :param path_to_archive: The path to the local archive
+        :return: The dict `local_files` that maps between local file and size in bytes
+        """
+        log.info("Generating local filelist for {}...".format(path_to_archive))
+        local_files = {}
+        for root, directories, filenames in os.walk(path_to_archive):
+            for filename in filenames:
+                full_path = os.path.join(root, filename)
+                local_size = os.path.getsize(full_path)
+                local_files[full_path] = int(local_size)
+
+        if not local_files:
+            raise ArteriaUsageException("Error when generating local filelist. No files found for {}".format(path_to_archive))
+
+        log.debug("Local files for the archive are {}".format(local_files))
+
+        return local_files
+
+    def get_files_to_reupload(self, local_files, uploaded_files):
```

Comment:
I'm wondering here. Is it correct for our use case to only base if we want to reupload on file size of? Or are there cases where the content of a file has changed, but size remains the same, but we still want to reupload it?

Reply:
It is not necessarily fully correct, but this is mainly to facilitate the "archive missing" workflow. I.e. files are actually missing. If the runfolder has been re-analyzed and/or one want to force a complete "reupload" then one should run the main "archive upload" workflow instead, which will create a new archive, and a new description tag (the unique identifier for an upload) on the PDC side, which will then be uploaded. I.e. that's not what I view as a reupload in this context.

Reply:
:+1: